### PR TITLE
SNIP-86: Sign-in-with-Starknet Standard

### PR DESCRIPTION
**Pull request by 0xYudhishthra**
Long messages can now be signed with this new standard using the Poseidon hash 🎉

The current SIWS standard introduces a lot of limitations when signing messages by enforcing a 31-char limit on the `StarknetDomain` and `Message` fields. This proposal suggests the removal of this limitation to allow for more flexibility in the use of the SIWS standard and increase adoption of the new standard for more dApps built on Starknet.

### CODE REVIEW

**Comment by JorikSchellekens** [comment]
Note: We are reworking the intro to this SNIP to be stand-alone instead of referencing the previous, closed pull request.

**Comment by 0xYudhishthra** [comment]
> Long messages can now be signed with this new standard using the Poseidon hash 🎉
> 
> The current SIWS standard introduces a lot of limitations when signing messages by enforcing a 31-char limit on the `StarknetDomain` and `Message` fields. This proposal suggests the removal of this limitation to allow for more flexibility in the use of the SIWS standard and increase adoption of the new standard for more dApps built on Starknet.

Made updates to this SNIP to be stand-alone and removed references to previous SNIPs, included detailed spec and implementation details for added context

**Comment by dor-starkware** [comment]
Hi @0xYudhishthra ,

Thank you for your submission of SNIP-86. To comply fully with SNIP-1 guidelines, please address the following:

**Rationale**: Include a section that describes the reasons for specific design choices and any alternatives that were considered.

**Backwards Compatibility**: Add a section that discusses any backward compatibility issues, even if there are none.

**Security Considerations**: While security aspects are discussed, please add a clearly labeled section titled "Security Considerations."

Once these changes are made, we can proceed further with the review process.

File: SNIPS/snip-86.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "type": "object",
+  "properties": {
+    "domain": {
+      "type": "object",
+      "properties": {
+        "chainId": {
+          "type": "string",
+          "enum": ["SN_SEPOLIA", "SN_MAIN"],
```
Comment (Eikix) [inline SNIPS/snip-86.md:58]: Why should we only limit it to two chains? There are a few appchains already (paradex for instance) and probably more soon; 
Moreover, you'd ideally want this standard to work on a local devnet too without using the devnet with testnet chainId imo

File: SNIPS/snip-86.md [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        },
+        "address": {
+          "type": "string",
+          "pattern": "^0x[a-fA-F0-9]{63,64}$",
+          "errorMessage": "Address must be a hexadecimal string with 66 characters, including the '0x' prefix"
+        },
+        "issuedAt": {
+          "type": "string",
+          "format": "date-time",
+          "errorMessage": "IssuedAt must be a valid date-time string"
```
Comment (Eikix) [inline SNIPS/snip-86.md:98]: worth it to precise "ISO8601" too? To have similarity with EIP-4361

File: SNIPS/snip-86.md [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+
+### Full Flow of Sign-In Process
+
+1. Challenge Preparation: A Starknet application (the challenger) prepares a sign-in request based on the specified schema. This request includes the domain and message objects.
+
+2. Signature Request: The user's wallet application displays the sign-in request, ensuring the user understands what they are signing.
+
+3. User Consent and Signature: If the user consents to the sign-in request, they sign the message using their Starknet account. This signature is then sent back to the challenger.
+
+4. Verification: The challenger uses an RPC call to a Starknet node to verify the signature against the user's account using the isValidSignature method. Due to account abstraction, the exact signature scheme doesn’t need to be known by the challenger; it only needs to ensure that the signature is valid for the given account.
```
Comment (Eikix) [inline SNIPS/snip-86.md:227]: why isValidSignature? This breaks snake case convention on Starknet

File: SNIPS/snip-86.md [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            "required": ["name", "type"],
+            "additionalProperties": false,
+            "errorMessage": "Items must include name and type"
+          },
+          "minItems": 6,
+          "maxItems": 8,
+          "uniqueItems": true,
+          "errorMessage": "Message must contain min 6-8 unique items "
+        },
+        "StarknetDomain": {
```
Comment (Eikix) [inline SNIPS/snip-86.md:180]: where is this used? it doesn't seem like you are typing "domain" property above

File: SNIPS/snip-86.md [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    },
+    "primaryType": {
+      "type": "string",
+      "const": "Message",
+      "errorMessage": "PrimaryType must be 'Message'"
+    },
+    "types": {
+      "type": "object",
+      "properties": {
+        "Message": {
```
Comment (Eikix) [inline SNIPS/snip-86.md:146]: where is this used? in primaryType? why not in "message" property?

**Comment by Eikix** [comment]
Adding an update here, I think I added comments too fast! I might have misunderstood how JSON schemas work. Apologies (can't resolve my comments)

**Comment by 0xYudhishthra** [comment]
> Adding an update here @0xYudhishthra, I think I added comments too fast! I might have misunderstood how JSON schemas work. Apologies (can't resolve my comments)

Hey @Eikix, no worries on that! Could you please let me know which comments you'd like resolved, so I know which ones to take a second look at? (just mention the comments in the numeric order, I'll help resolve it)

**Comment by Eikix** [comment]
> > Adding an update here @0xYudhishthra, I think I added comments too fast! I might have misunderstood how JSON schemas work. Apologies (can't resolve my comments)
> 
> Hey @Eikix, no worries on that! Could you please let me know which comments you'd like resolved, so I know which ones to take a second look at? (just mention the comments in the numeric order, e.g. 1,2,3... I'll help resolve it)

Ok now that I'm a SNIP maintainer, I could resolve comments! So all good now.

File: SNIPS/snip-86.md (hunk 3 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:227]: This is a typo, will change it to `is_valid_signature`

File: SNIPS/snip-86.md [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+- notBefore (optional): The earliest timestamp at which the authentication request can be considered valid, preventing its use before a certain time.
+
+### Sign-in Schema
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "type": "object",
+  "properties": {
+    "domain": {
```
Comment (Eikix) [inline SNIPS/snip-86.md:53]: I guess my initial question was: "How do you like "domain" key, and the custom defined type "StarknetDomain"?

Why don't you have something like:

```json
"properties": {
    "domain": {
      "$ref": "#/definitions/StarknetDomain",
      "errorMessage": "Domain must include chainId, name, version, revision"
    },
    "message": {
      "$ref": "#/definitions/Message",
      "errorMessage": "Message must include address, issuedAt, nonce, statement, uri, version"
    },
    ```

File: SNIPS/snip-86.md (hunk 1 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:73]: Regarding this, the challenge would be to identify a standard format in which we can generalize over chainIDs such as EVM-chains where it starts with 1. I noticed that such a method does not exist yet with Starknet where it can be identified with an integer (might be wrong here), perhaps for a start we can allow these two first?

File: SNIPS/snip-86.md (hunk 2 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:113]: I've included a "Message Fields" section to explain these fields in detail, similar to EIP-4361, looks good?

File: SNIPS/snip-86.md (hunk 6 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:68]: I didn't seem to put much thought into this yet since it was able to compile properly in the demo codebase, but let me look into this to see if a `$ref` is needed

**Comment by 0xYudhishthra** [comment]
> Hi @0xYudhishthra ,
> 
> Thank you for your submission of SNIP-86. To comply fully with SNIP-1 guidelines, please address the following:
> 
> **Rationale**: Include a section that describes the reasons for specific design choices and any alternatives that were considered.
> 
> **Backwards Compatibility**: Add a section that discusses any backward compatibility issues, even if there are none.
> 
> **Security Considerations**: While security aspects are discussed, please add a clearly labeled section titled "Security Considerations."
> 
> Once these changes are made, we can proceed further with the review process.

Hey @dor-starkware, I've included these sections...hope it's in order!

File: SNIPS/snip-86.md [hunk 7]
```diff
@@ -0,0 +1,601 @@
...
+requires: SNIP-12
+---
+
+## Summary
+
+The Sign-in with Starknet Standard (SIWS) enhances the flexibility and security of user authentication across Starknet by leveraging account abstraction and extended data fields. This standard supports a wider range of authentication scenarios, thereby improving user experience and expanding application development possibilities on Starknet.
+
+## Abstract
+
+The proposed SIWS standard introduces an advanced framework for authentication that enables signing of messages of any length, removing character limitations. This enhancement is made possible through the integration of the Poseidon hash function, which supports larger data sizes. The standard aims to facilitate broader adoption across decentralized applications by offering enhanced flexibility and robust security features.
```
Comment (gaetbout) [inline SNIPS/snip-86.md:19]: Could you kindly point me out where you saw that:
> the Poseidon hash function, which supports larger data sizes.
?

Comment (0xYudhishthra) [reply SNIPS/snip-86.md:19]: Perhaps I can reword it, but what I understood about how the Poseidon hash works is that it has significantly fewer constraints per message bit than Pedersen hash therefore making it suitable to sign larger **message** sizes

Also, would like to highlight on SNIP-12 (https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md) which demonstrates this. I also emphasised on sticking to revision 1, cause of this too -> https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md?plain=1#L196

Comment (gaetbout) [reply SNIPS/snip-86.md:19]: Thing is that Poseidon was pushed quite a lot due to the fact that it is/will be "more efficient" but apparently with STWO (coming early 2025) Pedersen might be the most efficient one in the end... 
This is an annoying situation 😞 

Yes it means that there might be a new version for SNIP-12 to use the most efficient hashing function.

File: SNIPS/snip-86.md [hunk 8]
```diff
@@ -0,0 +1,601 @@
...
+- `message` REQUIRED. An object containing the core components of the SIWS message.
+
+  - `version` REQUIRED. The version of the SIWS message. Its value MUST be a string in the format x.y.z and cannot exceed 31 characters.
+  - `address` REQUIRED. The Starknet address performing the signing. Its value MUST be a hexadecimal string with 66 characters, including the '0x' prefix.
+  - `statement` REQUIRED. A human-readable ASCII assertion that the user will sign.
+  - `uri` REQUIRED. An RFC 3986 URI referring to the resource that is the subject of the signing.
+  - `nonce` REQUIRED. A random string used to prevent replay attacks, at least 8 alphanumeric characters and no more than 31 characters.
+  - `issuedAt` REQUIRED. The time when the message was generated. Its value MUST be an ISO 8601 datetime string.
+  - `expirationTime` OPTIONAL. The time when the signed authentication message is no longer valid. Its value MUST be an ISO 8601 datetime string.
+  - `notBefore` OPTIONAL. The time when the signed authentication message will become valid. Its value MUST be an ISO 8601 datetime string.
```
Comment (gaetbout) [inline SNIPS/snip-86.md:253]: To show that those fields are somehow related, I'd rename them.
I believe the ERC4337 naming is better: `validUntil` and `validAfter`

File: SNIPS/snip-86.md [hunk 9]
```diff
@@ -0,0 +1,601 @@
...
+  - `revision` REQUIRED. The Typed Data revision. Its value MUST be a string and MUST be "1" for this specification.
+
+- `message` REQUIRED. An object containing the core components of the SIWS message.
+
+  - `version` REQUIRED. The version of the SIWS message. Its value MUST be a string in the format x.y.z and cannot exceed 31 characters.
+  - `address` REQUIRED. The Starknet address performing the signing. Its value MUST be a hexadecimal string with 66 characters, including the '0x' prefix.
+  - `statement` REQUIRED. A human-readable ASCII assertion that the user will sign.
+  - `uri` REQUIRED. An RFC 3986 URI referring to the resource that is the subject of the signing.
+  - `nonce` REQUIRED. A random string used to prevent replay attacks, at least 8 alphanumeric characters and no more than 31 characters.
+  - `issuedAt` REQUIRED. The time when the message was generated. Its value MUST be an ISO 8601 datetime string.
```
Comment (gaetbout) [inline SNIPS/snip-86.md:251]: Can you elaborate on the purpose of this field?

File: SNIPS/snip-86.md [hunk 10]
```diff
@@ -0,0 +1,601 @@
...
+2. **Generating the Signature**:
+
+   - **Account Abstraction**: Signature logic is defined by the account’s smart contract.
+   - **Signing Process**: User’s private key signs the hashed message securely within Starknet.
+
+#### Verifying Messages with Starknet Accounts
+
+1. **Verification Method for EOAs**:
+
+   - **EIP-712**: Verification involves reconstructing the signed message and comparing it against the provided signature.
```
Comment (gaetbout) [inline SNIPS/snip-86.md:402]: This is referencing EIP which aren't compatible with Starknet. Wouldn't it be better to reference the SNIPs?
I believe this is SNIP-12 and EIP-4361 is SNIP-6

File: SNIPS/snip-86.md (hunk 9 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:251]: To confirm, you mean the `issuedAt` field?

Comment (gaetbout) [reply SNIPS/snip-86.md:251]: Yes, sorry 😅

File: SNIPS/snip-86.md (hunk 10 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:402]: Agreed, included in 86b90d8

File: SNIPS/snip-86.md (hunk 9 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:251]: 2 considerations I had in mind are that:- 

- it can help prevent replay attacks to ensure the message is reused within an appropriate timeframe
- can be used together with `validUntil` and/or `validAfter` to make sure that the message is only valid for a certain duration

File: SNIPS/snip-86.md (hunk 7 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:19]: Yikes...I remember the SNIP also mentioning that support for strings longer than 31 characters would be introduced in revision `1` here -> https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md?plain=1#L353

Though it seems like this is already supported in Starknet.js after version `6.8.0`

File: SNIPS/snip-86.md [hunk 11]
```diff
@@ -0,0 +1,601 @@
...
+
+#### Message Fields
+
+This specification defines the following SIWS Message fields that can be parsed from a SIWS Message by following the rules in [Structured JSON Message Format](#structured-json-message-format).
+
+- `domain` REQUIRED. An object containing information about the domain requesting the signing.
+
+  - `chainId` REQUIRED. The Starknet network identifier. Its value MUST be either "SN_SEPOLIA" or "SN_MAIN".
+  - `name` REQUIRED. The name of the application requesting the signing.
+  - `version` REQUIRED. The version of the application. Its value MUST be a string in the format x.y.z and cannot exceed 31 characters.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:241]: i'd hardcode it to the version of SNIP-86 used

File: SNIPS/snip-86.md [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+```
+
+#### Message Fields
+
+This specification defines the following SIWS Message fields that can be parsed from a SIWS Message by following the rules in [Structured JSON Message Format](#structured-json-message-format).
+
+- `domain` REQUIRED. An object containing information about the domain requesting the signing.
+
+  - `chainId` REQUIRED. The Starknet network identifier. Its value MUST be either "SN_SEPOLIA" or "SN_MAIN".
+  - `name` REQUIRED. The name of the application requesting the signing.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:240]: IMO the domain name should be alway the same so wallets can recognize it and display it in a nice way, the app url can be part of the message, app name doesn’t need to be signed

File: SNIPS/snip-86.md [hunk 13]
```diff
@@ -0,0 +1,601 @@
...
+}
+```
+
+#### Message Fields
+
+This specification defines the following SIWS Message fields that can be parsed from a SIWS Message by following the rules in [Structured JSON Message Format](#structured-json-message-format).
+
+- `domain` REQUIRED. An object containing information about the domain requesting the signing.
+
+  - `chainId` REQUIRED. The Starknet network identifier. Its value MUST be either "SN_SEPOLIA" or "SN_MAIN".
```
Comment (sgc-code) [inline SNIPS/snip-86.md:239]: what's the reason to restrict to this 2 networks only?

File: SNIPS/snip-86.md [hunk 14]
```diff
@@ -0,0 +1,601 @@
...
+  - `name` REQUIRED. The name of the application requesting the signing.
+  - `version` REQUIRED. The version of the application. Its value MUST be a string in the format x.y.z and cannot exceed 31 characters.
+  - `revision` REQUIRED. The Typed Data revision. Its value MUST be a string and MUST be "1" for this specification.
+
+- `message` REQUIRED. An object containing the core components of the SIWS message.
+
+  - `version` REQUIRED. The version of the SIWS message. Its value MUST be a string in the format x.y.z and cannot exceed 31 characters.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:246]: IMO this field is redundant if we specify the version in de domain

File: SNIPS/snip-86.md [hunk 15]
```diff
@@ -0,0 +1,601 @@
...
+  - `address` REQUIRED. The Starknet address performing the signing. Its value MUST be a hexadecimal string with 66 characters, including the '0x' prefix.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:247]: this field is redundant as SNIP12 signature always include the signing address implicitly

File: SNIPS/snip-86.md [hunk 16]
```diff
@@ -0,0 +1,601 @@
...
+  - `statement` REQUIRED. A human-readable ASCII assertion that the user will sign.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:248]: i'd replace it with a felt : `challenge` to simplify the UI

File: SNIPS/snip-86.md [hunk 17]
```diff
@@ -0,0 +1,601 @@
...
+  - `uri` REQUIRED. An RFC 3986 URI referring to the resource that is the subject of the signing.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:249]: why is this needed?

File: SNIPS/snip-86.md [hunk 18]
```diff
@@ -0,0 +1,601 @@
...
+  - `nonce` REQUIRED. A random string used to prevent replay attacks, at least 8 alphanumeric characters and no more than 31 characters.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:250]: i'd enforce using a felt252, ideal size for security and efficienty, and it's not supposed to be prominent in the UI anyways

File: SNIPS/snip-86.md (hunk 9 above)
Comment (sgc-code) [reply SNIPS/snip-86.md:251]: IMO the field is not needed, the nonce will be enough to prevent replay attacks

File: SNIPS/snip-86.md [hunk 19]
```diff
@@ -0,0 +1,601 @@
...
+  - `issuedAt` REQUIRED. The time when the message was generated. Its value MUST be an ISO 8601 datetime string.
+  - `validUntil` OPTIONAL. The time when the signed authentication message is no longer valid. Its value MUST be an ISO 8601 datetime string.
+  - `validAfter` OPTIONAL. The time when the signed authentication message will become valid. Its value MUST be an ISO 8601 datetime string.
```
Comment (sgc-code) [inline SNIPS/snip-86.md:253]: IMO not needed, dapps can include this info in the challenge (`nonce`) and perform the logic they need while keeping this SNIP simple

File: SNIPS/snip-86.md [hunk 20]
```diff
@@ -0,0 +1,601 @@
...
+  "primaryType": "Message",
+  "types": {
+    "Message": [
+      { "name": "version", "type": "string" },
+      { "name": "address", "type": "felt" },
+      { "name": "statement", "type": "string" },
+      { "name": "uri", "type": "string" },
+      { "name": "nonce", "type": "string" },
+      { "name": "issuedAt", "type": "string" },
+      { "name": "validUntil", "type": "string" }
```
Comment (sgc-code) [inline SNIPS/snip-86.md:292]: i'd rather remove this fields but if they are needed we should use `timestamp` from SNIP12 over strings

File: SNIPS/snip-86.md [hunk 21]
```diff
@@ -0,0 +1,601 @@
...
+    "uri": "https://example.com",
+    "nonce": "32891756",
+    "issuedAt": "2024-05-24T12:34:56Z",
+    "validUntil": "2024-05-24T13:34:56Z"
+  },
+  "primaryType": "Message",
+  "types": {
+    "Message": [
+      { "name": "version", "type": "string" },
+      { "name": "address", "type": "felt" },
```
Comment (sgc-code) [inline SNIPS/snip-86.md:287]: field is not needed IMO, but in case it's needed the type `ContractAddress` is a better fit

File: SNIPS/snip-86.md [hunk 22]
```diff
@@ -0,0 +1,601 @@
...
+      { "name": "statement", "type": "string" },
+      { "name": "uri", "type": "string" },
+      { "name": "nonce", "type": "string" },
+      { "name": "issuedAt", "type": "string" },
```
Comment (sgc-code) [inline SNIPS/snip-86.md:291]: SNIP12 encourages human readable names, like "Issued At" and "Valid Until"

File: SNIPS/snip-86.md (hunk 9 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:251]: Open to ideas on this, perhaps the others can chime in here to see if its really needed

File: SNIPS/snip-86.md (hunk 14 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:246]: Ok, will remove this field

File: SNIPS/snip-86.md (hunk 15 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:247]: Removing this too

File: SNIPS/snip-86.md (hunk 16 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:248]: Perhaps, for implementers, a generic statement can be introduced in contrast to customizable ones

File: SNIPS/snip-86.md (hunk 17 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:249]: Explain better on why its needed, is it website domain, if so how is it specified? How to check if this is valid? What kind of checks can be done by the wallet to ensure this is correct.

TL;DR it's needed, just be more specific

File: SNIPS/snip-86.md (hunk 18 above)
Comment (0xYudhishthra) [reply SNIPS/snip-86.md:250]: Will change this

Comment (0xYudhishthra) [reply SNIPS/snip-86.md:250]: Nonce will be visible initially in the UI, over time, will be hidden in wallet signing UI
